### ISSUE TITLE
Building failure: ISO C++17 does not allow 'register' storage class specifier [-Wregister]

### ISSUE THREAD

**Issue by wilstoff** [doxygen/doxygen#10649]
**Describe the bug**
When running `cmake/make` with
```
cmake -G "Unix Makefiles" -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/usr -Wno-dev
```
it seems to generate a bunch of code with the register keyword that seems to be deprecated.
Which errors out when building:
```
/build/bstoffel/rpmbuild/BUILD/doxygen-1.10.0/generated_src/mscgen_lexer.cpp:899:2: error: ISO C++17 does not allow 'register' storage class specifier [-Wregister]
        register yy_state_type yy_current_state;
        ^~~~~~~~~
/build/bstoffel/rpmbuild/BUILD/doxygen-1.10.0/generated_src/mscgen_lexer.cpp:900:2: error: ISO C++17 does not allow 'register' storage class specifier [-Wregister]
        register char *yy_cp, *yy_bp;
...
```

**Expected behavior**
I expect the code to compile appropriately

**Version**
Building version 1.10.0

**Additional context**
Running on centos 7 machine with clang 10.0.0 installed.
I have tried adjusting the C++ version to 14/11 and still has the same errors.
I'm sure there is an option I'm missing or can alter some environmental variable in order for it to generate valid C++17 code.

**Comment by albert-github** [comment]
When I build doxygen I see in the resulting `mscgen_lexer.cpp` file:
```
/** The main scanner function which does all the work.
 */
YY_DECL
{
        yy_state_type yy_current_state;
        char *yy_cp, *yy_bp;
        int yy_act;
```

I'm using flex version 2.6.4 locally.
On GitHub actions we also use the CLang compiler, version 10.0.0 with flex 2.6.4 (see e.g. https://github.com/doxygen/doxygen/actions/runs/7877930596/job/21495098205) and here I don't see any problems there.


- which version of flex are you using? or are you using `lex`?
- can you post the resulting `mscgen_lexer.cpp`  file in a, compressed, tar or zip file attached to the issue?

Looking in the file `CMakeLists.txt` we see for the MacOs Darwin version:
```
    set(CMAKE_CXX_FLAGS "-Wno-deprecated-register -mmacosx-version-min=${MACOS_VERSION_MIN} ${CMAKE_CXX_FLAGS}")
    set(CMAKE_C_FLAGS "-Wno-deprecated-register -mmacosx-version-min=${MACOS_VERSION_MIN} ${CMAKE_C_FLAGS}")
```

I see that CLang has also an option:
> -Wdeprecated, -Wno-deprecated[¶](https://releases.llvm.org/10.0.0/tools/clang/docs/ClangCommandLineReference.html#cmdoption-clang-wdeprecated)
>
> Enable warnings for deprecated constructs and define __DEPRECATED

so maybe something like
```
    set(CMAKE_CXX_FLAGS "-Wno-deprecated ${CMAKE_CXX_FLAGS}")
    set(CMAKE_C_FLAGS "-Wno-deprecated ${CMAKE_C_FLAGS}")
```
should be added, conditionally, to the `CMakeLists.txt` file, but first we have to see what the version of flex etc. is that is used.

**Comment by matthewjmiller1** [comment]
I am seeing this too in a legacy build environment. It looks like its flex is 2.5.39. Using `-Wno-deprecated` didn't seem to help.

**Comment by matthewjmiller1** [comment]
Here is the resulting `mscgen_lexer.cpp` file (I replaced the real workspace path in some of the `#line` lines with `/workspace`):
[mscgen_lexer.cpp.tar.gz](https://github.com/user-attachments/files/16822727/mscgen_lexer.cpp.tar.gz)

**Comment by albert-github** [comment]
@matthewjmiller1 
Looks like to be a problem with the a bit older flex versions.
- Do you see this problem only with `mscgen_lexer.cpp` / `mscgen_lexer.l` or also with other `.l` files and or `.cpp` / `.c` files?
- You wrote about `-Wno-deprecated` but in my comment (https://github.com/doxygen/doxygen/issues/10649#issuecomment-1941271533) I wrote about `-Wno-deprecated-register` (though I don't see the `-Wno-deprecated-register`in the gcc manual page I have) what did you use?
- What happens when you place `#define register` as line 38 in the file `mscgen_lexer.l`?

**Comment by matthewjmiller1** [comment]
@albert-github 
It seems to be other files too, but not sure how many b/c the build bails after several errors. `xml.cpp` failed after `mscgen_lexer.cpp`. I'd guess there may be others after that.

I tried `-Wno-deprecated-register` and that didn't help.

I tried adding `#define register` to line 38 of `mscgen_lexer.l` and that did seem to get past the `mscgen_lexer.cpp` issue, but then I hit the `xml.cpp` issue.

I wrote an Ubuntu 24.04 Dockerfile that you can use to repro this:

[doxygen_cmake.tar.gz](https://github.com/user-attachments/files/16826126/doxygen_cmake.tar.gz)

There is a `README.md` in there that tells you the commands to use. It creates two clones of doxygen: one to test `-Wno-deprecated-register` and the other to test the modified `mscgen_lexer.l` file.

**Comment by albert-github** [comment]
It is like I expected
- with `#define resister` the problem disappears for that file
- the problem is present in other wfiles as well

For the solution it would be good to now whether the problem only occurs in files generated from `.l` files or also on other places. I see the word `register` also in `deps/libmscgen/gd.c`.
You cans see this by means of `make -i` which will compile all files and issue the warnings but won't stop compiling. In this way  you can see all the files that will give problems.

My expectations are:
- `deps/libmscgen/gd.c` won't give problems
- all `.l` file will give a compilation warning
- no further files will give a problem.

So the completer list probably would be:
- `config_doxyw.cpp`
- `mscgen_lexer.cpp`
- `xml.cpp`
- `code.cpp`
- `commentcnv.cpp`
- `commentscan.cpp`
- `configimpl.cpp`
- `constexp.cpp`
- `declinfo.cpp`
- `defargs.cpp`
- `doctokenizer.cpp`
- `fortrancode.cpp`
- `fortranscanner.cpp`
- `lexcode.cpp`
- `lexscanner.cpp`
- `pre.cpp`
- `pycode.cpp`
- `pyscanner.cpp`
- `scanner.cpp`
- `sqlcode.cpp`
- `vhdlcode.cpp`
- `xmlcode.cpp`

@matthewjmiller1 can you confirm this?

As additional information he c++ 2023 standard stes:
> The register keyword is unused but is reserved for future use

For the first 3 files I think the way to go would be to add the line `#define register` in the file
Fort the other files (all on the `src` directory I think the way to go is to add the line `#define register` to the file `util.h` or `doxygen.h`

Another way would be to add the `-Dregister` to the `CMAKE_CXX_FLAGS` in case of a flex versions < 2.6.0
(In the NEWS file of the https://github.com/westes/flex repository I see 
```
* version 2.6.0 released 2015-11-17
...
*** Removed deprecated 'register' storage class specifier
```

@doxygen what do you think, which way to choose?

**Comment by matthewjmiller1** [comment]
@albert-github 
This is the list of files I saw in my build with `make -i`:
```
xmlcode_intermediate.cpp
vhdlcode_intermediate.cpp
sqlcode_intermediate.cpp
scanner_intermediate.cpp
pyscanner_intermediate.cpp
pycode_intermediate.cpp
pre_intermediate.cpp
lexscanner_intermediate.cpp
lexcode_intermediate.cpp
fortranscanner_intermediate.cpp
fortrancode_intermediate.cpp
doctokenizer_intermediate.cpp
defargs_intermediate.cpp
declinfo_intermediate.cpp
constexp_intermediate.cpp
commentscan_intermediate.cpp
commentcnv_intermediate.cpp
code_intermediate.cpp
configimpl_intermediate.cpp
xml.cpp
mscgen_lexer.cpp
```

Adding `-DCMAKE_CXX_FLAGS="-Dregister=''"` to the cmake CLI seemed to work for those errors (I saw linking errors later, not sure if they were related).

Adding `-DCMAKE_CXX_FLAGS="-Dregister"` didn't, giving errors like:

```
/git_repos/doxygen_no_deprecated/cmake_build/generated_src/mscgen_lexer.cpp:1509:10: error: expected ';' after expression
 1509 |         register int number_to_move, i;
      |                 ^
      |                 ;
```

The Dockerfile tarball from https://github.com/doxygen/doxygen/issues/10649#issuecomment-2323050452 can be used to test all these permutations.

**Comment by albert-github** [comment]
Perfect, exactly the list I expected (actually I expected one more file: `config_doxyw_intermediate.cpp`, but you didn't get this one as you didn't enable the doxywizard).

`-DCMAKE_CXX_FLAGS="-Dregister=''"` wont work as the line:
```
register int number_to_move, i;
```
would be translated into:
```
'' int number_to_move, i;
```
and this is not valid C / C++

I thought it would be easy to fix but MSVC gives some problems:
```

MSVC made some problems:
```
D:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\include\xkeycheck.h(221): warning C4005: 'register': macro redefinition
...\doxygen\deps\spdlog\src\spdlog.cpp: note: see previous definition of 'register'
D:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\include\xkeycheck.h(260): fatal error C1189: #error:  The C++ Standard Library forbids macroizing keywords. Enable warning C4005 to find the forbidden macro.
```
and in https://learn.microsoft.com/en-us/cpp/porting/visual-cpp-change-history-2003-2015?view=msvc-170I found:
>C++11 17.6.4.3.1 [macro.names]/2 forbids macro replacement of keywords when C++ Standard Library headers are included. The headers now emit compiler errors if they detect macro-replaced keywords. (Defining _ALLOW_KEYWORD_MACROS allows such code to compile, but we strongly discourage that usage.)


I've just pushed a proposed patch, pull request #11113, though I would advise also to update the flex version to 2.6.4 (which anyway is also already from 2017, but is the newest release available)

**Comment by albert-github** [comment]
Code has been integrated in master on GitHub (please don't close the issue as this will be done at the moment of an official release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.13.0.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
